tally/src/scope_impl.cc

```diff
@@ -0,0 +1,254 @@
+  std::lock_guard<std::mutex> lock(this->timers_mutex_);
+
+  // Since the timer reports metrics itself it must be initialized with the
+  // fully qualified name.
+  auto entry =
+      this->timers_.insert(std::pair<std::string, std::shared_ptr<TimerImpl>>(
+          name, TimerImpl::New(FullyQualifiedName(name), tags_, reporter_)));
+  return entry.first->second;
+}
+
+std::shared_ptr<tally::Histogram> ScopeImpl::Histogram(const std::string &name,
+                                                       const Buckets &buckets) {
+  std::lock_guard<std::mutex> lock(this->histograms_mutex_);
+
+  auto entry = this->histograms_.insert(
+      std::pair<std::string, std::shared_ptr<HistogramImpl>>(
+          name, HistogramImpl::New(buckets)));
+  return entry.first->second;
+}
+
+std::shared_ptr<tally::Scope> ScopeImpl::SubScope(const std::string &name) {
+  return SubScope(FullyQualifiedName(name),
+                  std::unordered_map<std::string, std::string>{});
+}
+
+std::shared_ptr<tally::Scope> ScopeImpl::Tagged(
+    const std::unordered_map<std::string, std::string> &tags) {
+  return SubScope(prefix_, tags);
+}
+
+std::unique_ptr<tally::Capabilities> ScopeImpl::Capabilities() {
+  if (reporter_ == nullptr) {
+    return std::unique_ptr<tally::Capabilities>(new CapableOf(false, false));
+  }
+  return reporter_->Capabilities();
+}
+
+std::shared_ptr<tally::Scope> ScopeImpl::SubScope(
+    const std::string &prefix,
+    const std::unordered_map<std::string, std::string> &tags) {
+  std::unordered_map<std::string, std::string> new_tags;
+
+  // Insert the new tags second as they take priority over the scope's tags.
+  for (auto const &tag : tags_) {
+    new_tags.insert(tag);
+  }
+
+  for (auto const &tag : tags) {
+    new_tags.insert(tag);
+  }
+
+  auto id = ScopeID(prefix, new_tags);
+
+  std::shared_ptr<Scope> scope = ScopeBuilder()
+                                     .reporter(reporter_)
+                                     .prefix(prefix)
+                                     .separator(separator_)
+                                     .tags(new_tags)
+                                     .Build();
+
+  std::lock_guard<std::mutex> lock(this->registry_mutex_);
+  auto entry =
+      this->registry_.insert(std::pair<std::string, std::shared_ptr<ScopeImpl>>(
+          id, std::shared_ptr<ScopeImpl>(
+                  std::dynamic_pointer_cast<ScopeImpl>(scope))));
+  return entry.first->second;
+}
+
+std::string ScopeImpl::FullyQualifiedName(const std::string &name) {
+  if (prefix_ == "") {
+    return name;
+  }
+
+  std::string str;
+  str.reserve(prefix_.length() + separator_.length() + name.length());
+  std::stringstream stream(str);
+  stream << prefix_ << separator_ << name;
+
+  return stream.str();
+}
+
+std::string ScopeImpl::ScopeID(
+    const std::string &prefix,
+    const std::unordered_map<std::string, std::string> &tags) {
+  std::vector<std::string> keys;
+  keys.reserve(tags.size());
```

Comment:
Good job on using `reserve` here. But you can do this whole thing with stl without using loops. As Scott Meyers writes in Effective STL, use algorithms instead of for loops where possible. It makes your loops more clear.

```
keys.reserve(tags.size());
std::transform(std::begin(tags), std::end(tags), std::back_inserter(keys), [](const std::pair<std::string, std::string>& tag) { return tag.first; })
```